bugsnag-android-core/src/main/java/com/bugsnag/android/ThreadState.kt

```diff
@@ -109,16 +115,41 @@ internal class ThreadState @Suppress("LongParameterList") constructor(
 
         // Keep the lowest ID threads (ordered). Anything after maxThreadCount is lost.
         // Note: We must ensure that currentThread is always present in the final list regardless.
-        val keepThreads = allThreads.sortedBy { it.id }.take(maxThreadCount)
-
-        val reportThreads = if (keepThreads.contains(currentThread)) {
-            keepThreads
-        } else {
-            // API 24/25 don't record the currentThread, so add it in manually
-            // https://issuetracker.google.com/issues/64122757
-            // currentThread may also have been removed if its ID occurred after maxThreadCount
-            keepThreads.take(max(maxThreadCount - 1, 0)).plus(currentThread).sortedBy { it.id }
-        }.map { toBugsnagThread(it) }.toMutableList()
+
+        val sortedThreads = allThreads.sortedBy { it.id }
+        val currentThreadIndex = sortedThreads.binarySearch(0, min(maxThreadCount, sortedThreads.size)) {
+            it.id.compareTo(currentThread.id)
+        }
+
+        // API 24/25 don't record the currentThread, so add it in manually
+        // https://issuetracker.google.com/issues/64122757
+        // currentThread may also have been removed if its ID occurred after maxThreadCount
+        val keepThreads = sortedThreads.take(
+            if (currentThreadIndex >= 0) maxThreadCount else max(maxThreadCount - 1, 0)
+        )
```

Comment:
Given the change in logic, I think we should enhance the comment to explain the `maxThreadCount - 1`:

```suggestion
        // API 24/25 don't record the currentThread, so add it in manually
        // https://issuetracker.google.com/issues/64122757
        // currentThread may also have been removed if its ID occurred after maxThreadCount
        // as such we may need to leave a space in new list for currentThread
        val keepThreads = sortedThreads.take(
            if (currentThreadIndex >= 0) maxThreadCount else max(maxThreadCount - 1, 0)
        )
```